examples/shaders/transion: Update Luma Burn

Oncorporation opened the pull request:
### Description
Adds transition color to luma-burn in both normal and inverted. This creates some dramatic effects

implement color to be consistent with sliding bars
add 2 techniques to separate the color calculations if unneeded
default color is a dramatic dark red

Oncorporation commented:
Previous pull request was too complex to resolve in github, I am more of a tfs guy. I eventually rebased and reset back so far the pull request got lost! freaking cherry picking merges!

SnopsisDev commented:
> 
> 
> LGTM. Don't know if the lack of a `.` for the float4 values will cause issues on non-Windows, still have to test that.

I am happy to just update that. This is your project and I want to be consistent

Xaymar commented:
You can rebase your changes onto master with `git rebase -i origin/master` (replace origin/master with the name of the remote). This puts you in an interactive git rebase mode, where you can select the state of certain commits. See the [git book entry about rewriting history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History), which explains it quite well. Or alternatively use a visual client like TortoiseGit, GitHub Desktop, SourceTree, GitKraken, etc

Xaymar commented:
Tested this out, this currently results in a harsh color transition at the end of it, which is definitely unwanted. The effect itself would look great once fixed though. 👍

SnopsisDev commented:
One thing that I noticed is that ColoredInverse and Colored are probably reversed. The harsh color change at the end is resolved in ColoredInverse as the color is added to the A scene and fades out. Colored is added to the B scene and fades in, then just goes away... "tada!"

They are inverses, but the gentile one is more expected. How would you feel about swapping these two?

Oncorporation commented:
I had posted from a consulting account, sorry...

One thing that I noticed is that ColoredInverse and Colored are probably reversed. One goes from light to dark, the other from dark to light.

The harsh color change at the end can be resolved by adjusting the alpha or matching the color of the next scene, but it is intentionally a "tada!"

We could add separate technique(s) or a "fade color" option to make it more gentile, but I made [this little video](https://www.twitch.tv/videos/694108432) to show you why I have it this way at the moment.

Oncorporation commented:
Changed how techniques are named. Now they express as 
Light to Dark or Dark to Light, 
representing how the luma is transitioned.

If they have coloration, the "Colored" name is appened to the description where it is inserted.
Light Colored to Dark
Light to Dark Colored
Dark Colored to Light
Dark to Light Colored

When the Coloration is on the right side, it blends out ...when on left it blends in and then "tada" at the end.
 
I think this will give you what you expect, and I have the option of "tada!"

Oncorporation commented:
I was able to successfully test on 4 different Windows machines, but all Windows 10 ...no Windows 7 yet.

Do you have a set of predefined OS's that you want tested against?

Xaymar commented:
Testing needs to be done on each major GPU vendor platform: Intel Haswell or newer, AMD GCN1 or newer, NVidia Maxwell or newer.

skeletonbow commented:
My main machine has AMD GCN1, and my laptop has Intel HD 630, and nVidia 1050Ti.  If there is a test build available I can give it a whirl on all 3.

Xaymar commented:
Test builds are available from CI: https://github.com/Xaymar/obs-StreamFX/actions/runs/190627840

Oncorporation commented:
Excellent Skeletonbow!  

I have tested so far on...

Windows 10.0.18363.959 LAPTOP
Kaby Lake GT2 GPU (Intel)
http://gpuz.techpowerup.com/20/08/04/9ab.png
Intel Core i7 8555U  Kaby Lake-U/Y

Windows 10.0.18363.959 Desktop
Navi 10 (AMD)
http://gpuz.techpowerup.com/20/08/04/ztu.png
Matisse (AMD)

Windows 10.0.19041.288 (Old Intel Desktop)
Clarkdale (Intel @2010)
http://gpuz.techpowerup.com/20/08/04/uta.png
Clarkdale (45nm Intel) lol

Putting together an Ubuntu 20.04 pen drive, but I do not have any recent nVidia cards

Xaymar commented on data/examples/shaders/transition/luma-burn.effect:
```diff
@@ -108,7 +117,7 @@ float4 PSDefault(VertData v_in) : TARGET {
 }
 
-technique Draw
+technique DrawDarkToLight
```
I did not notice this at first, but changing the name here would break existing configurations. Please keep both the original and the new name for it.

Oncorporation replied:
This and the next, I put at the beginning of the techniques to ensure they listed first. It doesn't line up as well for code review, but I think users will make more sense of it.

Xaymar commented on data/examples/shaders/transition/luma-burn.effect:
```diff
@@ -145,11 +154,176 @@ float4 PSInverse(VertData v_in) : TARGET {
 }
 
-technique DrawInverse
+technique DrawLightToDark
```
Same here.

Oncorporation replied:
I put at the beginning of the techniques

Oncorporation commented:
I could not for the life of me make the constant float3x3 to instantiate. 
There was a request for a required package in Ubuntu.  With OBS Studio and StreamFX installed, the additional package was  worked around.

Xaymar commented:
I'm very confused by what happened to the file, did you commit with the wrong line endings to make it suddenly explode to 2x the changes? Also what required package?

Oncorporation commented:
I had to refactor it to accommodate a lack of https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_shading_language_420pack.txt in Ubuntu.

It kept telling me that I could not do C style instantiations, so I took them out.  If that package is available to add , I think it needs a reference on the CPU side.

This was on a fresh instance on Ubuntu with the installation packages installed from the building documentation.

Oncorporation commented:
... if you have a solution, I don't want to refactor this either. Although, this works in Ubuntu and Windows

Xaymar commented:
Adding that extension to OpenGL would be an upstream change in OBS, not exactly something I can do in StreamFX yet. While I do plan to move the entire plugin over to Vulkan (MoltenVK on Mac), it is not quite there yet and my understanding of Vulkan is still a bit lacking right now.

Comparing the file on a linux machine reveals the addition of a CR character infront of the LF character. Please fix this, all the files in the repo use LF, not CRLF.

Oncorporation commented:
Funny, I was working on Ubuntu in Sublime Text 3, It didn't occur to me that it was using the wrong line endings.

Oncorporation commented:
Ok, new version tested in 
Windows 10.0.18363.959 LAPTOP
Kaby Lake GT2 GPU (Intel)
http://gpuz.techpowerup.com/20/08/04/9ab.png
Intel Core i7 8555U Kaby Lake-U/Y

Windows 10.0.18363.959 & Windows 10.0.19041.388 (Desktop )
Navi 10 (AMD)
http://gpuz.techpowerup.com/20/08/04/ztu.png
Matisse (AMD)

Windows 10.0.19041.288 (Old Intel Desktop)
Clarkdale (Intel @2010)
http://gpuz.techpowerup.com/20/08/04/uta.png
Clarkdale (45nm Intel) lol

Ubuntu 20.04
on intel with AMD 5600XT ( will fill in details)

Xaymar commented:
Got around to testing, couple of things that I found:

1. The forced pop-in effect just doesn't look right, and should be made optional through some sort of smoothness slider.
2. You don't have to prefix all technique names with `Draw`, and you can actually use spaces in them as long as they are surrounded by `""` - at least if the OBS shader parser supports it.

I also found a bug in my original code, enabling the exponential luminosity option inverts the normal draw method.

Oncorporation commented:
in progress...
1. Added an easing function to the colored techniques.  (Best results of 3 versions)
1a. Updated default values to make things more smooth.
2. Renamed the techniques

Oncorporation commented:
Easing function has now been successfully tested in Ubuntu! No changes needed
